sdk/core/System.ClientModel/src/Internal/SSE/ServerSentEventReader.cs

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Internal;
+
+/// <summary>
+/// An SSE event reader that reads lines from an SSE stream and composes them
+/// into SSE events.
+/// See SSE specification: https://html.spec.whatwg.org/multipage/server-sent-events.html
+/// </summary>
+internal sealed class ServerSentEventReader
+{
+    private readonly Stream? _stream;
+    private readonly StreamReader? _reader;
+
+    public ServerSentEventReader(Stream stream)
+    {
+        // Creator of the reader has responsibility for disposing the stream
+        // passed to the reader constructor.
+
+        _stream = stream;
+        _reader = new StreamReader(stream);
```

Comment:
Should we check for null on `stream` and validate that `_reader` isn't null after constructing the `StreamReader`? It might make this easier to debug than to wait until `TryGetNextEvent` throws sometime later.

Reply:
Good catch.  I'll check the argument - I don't think the _reader constructor can return null?